x-pack/plugins/ml/public/application/components/job_spaces_list/job_spaces_list.tsx

```diff
@@ -4,20 +4,64 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-import React, { FC } from 'react';
+import React, { FC, useState, useEffect } from 'react';
 
-import { EuiFlexGroup, EuiFlexItem, EuiBadge } from '@elastic/eui';
+import { EuiFlexGroup, EuiFlexItem, EuiButtonEmpty } from '@elastic/eui';
+import { JobSpacesFlyout } from '../job_spaces_selector';
+import { JobType } from '../../../../common/types/saved_objects';
+import { useSpacesContext } from '../../contexts/spaces';
+import { Space, SpaceAvatar } from '../../../../../spaces/public';
+
+export const ALL_SPACES_ID = '*';
 
 interface Props {
-  spaces: string[];
+  spaceIds: string[];
+  jobId: string;
+  jobType: JobType;
+  refresh(): void;
+}
+
+function filterUnknownSpaces(ids: string[]) {
+  return ids.filter((id) => id !== '?');
 }
 
-export const JobSpacesList: FC<Props> = ({ spaces }) => (
-  <EuiFlexGroup wrap responsive={false} gutterSize="xs">
-    {spaces.map((space) => (
-      <EuiFlexItem grow={false} key={space}>
-        <EuiBadge color={'hollow'}>{space}</EuiBadge>
-      </EuiFlexItem>
-    ))}
-  </EuiFlexGroup>
-);
+export const JobSpacesList: FC<Props> = ({ spaceIds, jobId, jobType, refresh }) => {
+  const { allSpaces } = useSpacesContext();
+
+  const [showFlyout, setShowFlyout] = useState(false);
+  const [spaces, setSpaces] = useState<Space[]>([]);
+
+  useEffect(() => {
+    const tempSpaces = spaceIds.includes(ALL_SPACES_ID)
+      ? [{ id: ALL_SPACES_ID, name: ALL_SPACES_ID, disabledFeatures: [], color: '#DDD' }]
+      : allSpaces.filter((s) => spaceIds.includes(s.id));
+    setSpaces(tempSpaces);
+  }, [spaceIds, allSpaces]);
+
+  function onClose() {
+    setShowFlyout(false);
+    refresh();
+  }
+
+  return (
+    <>
+      <EuiButtonEmpty onClick={() => setShowFlyout(true)} style={{ height: 'auto' }}>
+        <EuiFlexGroup wrap responsive={false} gutterSize="xs">
+          {spaces.map((space) => (
+            <EuiFlexItem grow={false} key={space.id}>
+              <SpaceAvatar space={space} size={'s'} />
```

Comment:
In most cases, the space avatar is rendered alongside a text description of the space, so we opted to make it "hidden" by default to avoid announcing it twice: https://github.com/elastic/kibana/issues/27748

Reply:
But if people don't know that `SpaceAvatar` is a purely visual component, this is the way more destructive default (i.e., it's better to double announce something by accident than not announce anything at all)

Reply:
That's a fair point. Up until now, the security team have been the only consumers of this component, and its original design didn't expect broad usage at the time. It probably makes sense to revisit this decision now that it's being used in more places by more teams, who don't know of its history